function/avro-import.py

```diff
@@ -0,0 +1,156 @@
+    except KeyError as ke:
+        return handle_exception(job_id, start_time, "Key Error: " + str(ke) + " in request " + str(request.json))
+
+    defaults = {'b64-decode-enums': True, 'prefix-object-ids': True}
+    request_options = request_json.get('options', {})
+    options = {**defaults, **request_options}
+
+    try:
+        avro = urllib.request.urlopen(url)
+    except urllib.error.URLError as ue:
+        return handle_exception(job_id, start_time, "URL Error - the url " + url + " is not valid.")
+
+    try:
+        reader = fastavro.reader(avro)
+        translation = translate(reader, options)
+        metadata = {
+            "namespace": workspace_namespace,
+            "name": workspace_name,
+            "userSubjectId": user_subject_id,
+            "userEmail": user_email,
+            "jobId": job_id,
+            "startTime": start_time,
+        }
+        metadata_json_str = str(json.dumps(metadata))
+        metadat_file_name = job_id + "/metadata.json"
+        upsert_json_str = str(json.dumps(translation))
+        upsert_file_name = job_id + "/upsert.json"
+        write_to_bucket(metadat_file_name, metadata_json_str)
+        write_to_bucket(upsert_file_name, upsert_json_str)
+    except Exception as e:
+        return handle_exception(job_id, start_time, "The following exception occurred: " + str(e))
+    except:
+        return handle_exception(job_id, start_time, "Something went wrong with the following request: " + str(request.json))
+
+
+def handle_exception(subdirectory_name, start_time, message):
+    error_file_name = subdirectory_name + "/error-" + start_time + ".txt"
+    write_to_bucket(error_file_name, message)
+    return message
+
+
+def write_upsert_to_bucket(job_id, content_string):
+    compressed_value = gzip.compress(bytes(content_string, 'utf-8'))
+    file_name = job_id + "/upsert.json"
+    write_to_bucket(file_name, compressed_value)
+
+
+def write_metadata_to_bucket(job_id, content_string):
+    file_name = job_id + "/metadata.json"
+    write_to_bucket(file_name, content_string)
+
+
+def write_to_bucket(file_name, content):
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket("avro-translated-json")
+    blob = bucket.blob(file_name)
+    blob.upload_from_string(content)
+
+
+def translate(reader, options=None):
+    if options is None:
+        options = {}
+    t = Translator(options)
+    return t.translate(reader)
+
+
+class Translator:
+    def __init__(self, options=None):
+        if options is None:
+            options = {}
+        defaults = {'b64-decode-enums': False, 'prefix-object-ids': False}
+        self.options = {**defaults, **options}
+
+    def translate(self, reader):
+        if reader is None:
+            return None
+
+        enums = _list_enums(reader.writer_schema)
+        results = [self._translate_record(record, enums)
+                   for record in reader if record['name'] != 'Metadata']
```

Comment:
dict comprehension! :heavy_check_mark:

Reply:
so you're expecting the `Metadata` key to come in from the user? is it required?

Reply:
I don't know - this was lifted in part from Brian's arrow service :)

Reply:
`Metadata` is a record with... well... metadata. It's not an entity. By the time we get here, we just want the entities so we skip `Metadata`.